### ISSUE TITLE
Strongly typed shape does not receive parameters correctly

### ISSUE THREAD

**Issue by ViRuSTriNiTy** [OrchardCMS/Orchard#5514]
Hi there,

i don't know whether this is an intended behavior or not. Here's a simple example:

Shape view model

    public class MyShape : Orchard.DisplayManagement.Shapes.Shape
    {
      public string Html { get; set; }
    }

Shape template (MyShape.cshtml)

    @model Dennree.UI.ViewModels.MyShape

    MyShape: @Model.Html

View displaying the shape

    @Display.MyShape(typeof(Dennree.UI.ViewModels.MyShape), Html: "MyShapeHtml")

Now, when `@Display...` is called `DefaultShapeFactory.Create()` creates the shape correctly and then adds a dynamic `Html` property with value `MyShapeHtml` to the shape. In debugger i can see that the declared property `Html` remains untouched (empty string) and this leads to the issue that `@Model.Html` in the shape template renders an empty string. I have to write `@Model.Properties["Html"]` in the shape template to get `MyShapeHtml` displayed.

Did i miss something or is writing `.Properties[]` the designated way in shape templates? In the later case i wonder why i even use a strong type for the model as `.Properties[]` is not really strongly typed in my book.

Currently i added an override of `TrySetIndex()` to the view model to get the property updated correctly.

So lonG

**Comment by nieve** [comment]
It's an old one, but just in case anyone tries to fight this undocumented (or un-googlable) functionality, if your MyShape class contains the following:
```cs
public MyShape() {
    Metadata = new ShapeMetadata {Type = "MyShape"};
}
```
You could now simply do this:
@Display(new MyShape{Html= "MyShapeHtml" })

HTH

**Comment by sebastienros** [comment]
Feel free to explain why you are create custom shape types, I have never seen anyone doing it, but not saying you shouldn't, very advanced usage.

**Comment by ViRuSTriNiTy** [comment]
I use shapes here and there to display certain non-content data (mostly it's very complex data).

For example i have a shape that adds an ajax progress bar, it receives currently no parameters but because it is a shape i can use it across all my modules with ease by simply putting `@Display(new UI.ViewModels.AjaxTimeoutProgressBar())` to any view in any module and it works.

But there are also very complex shapes like the `Table` shape which renders a fully blown table. Amongst other features the table provides an *area* to render filter options (`TableFilter` shape), caption ('TableCaption' shape), rows (`TableRow` shape), item pickers, paging buttons or a page size selection.

Now if i want to display a table i simply put `@Display(new UI.ViewModels.Table {Parameter1 = ...})` to a view and it will render the `Table` shape with the parameters given. If i want to adjust the table to the needs of a module, lets say i need different filter options, then i simply put a customized shape template named `TableFilter.cshtml` to the module and the table would now display different filter options. There is also an advanced use of `IShapeTableProvider` where i add a shape alternate for the `TableRow` shape based on a certain condition like "what display type is selected?". The table then renders the rows differently for display type A or display type B.

So to sum it up, i have a `UI` module which provides shapes for rendering often used UI elements and with the power of Orchards shape system it is possible to override parts of each shape or the whole shape itself to customize each shape according to the current need.

**Comment by nieve** [comment]
As for me, I have inherited an Orchard project, and being unfamiliar with Orchard, I was simply looking for a way to render partial views the Orchard way. Since I'd rather not work with dynamics, I was trying to figure out how it will be possible to combine a shape with a strongly typed model, et voili voilà.
I'd be happy to know if there is a simpler way to do it in Orchard. This is one of the most basic mechanisms in MVC and it's somewhat frustrating not knowing what is the common approach it.

My preference to steer clear of dynamics brings up the question of benchmarking: @ViRuSTriNiTy did you use a benchmarking framework such as [BenchmarkDotNet](https://github.com/PerfDotNet/BenchmarkDotNet) or did you do it manually?

**Comment by ViRuSTriNiTy** [comment]
@nieve I got all my knowledge about shapes from the following video on youtube:

https://www.youtube.com/watch?v=gKLjtCIs4GU, How to use strongly typed shapes - 18:46﻿

And there is also an article by @pszmyd (see http://www.szmyd.com.pl/blog/using-shapes-as-html-helpers-in-orchard) that described the internals a bit in detail.

And no, i didn't use a benchmark framework, i used the google chrome console (network tab) to measure response times.

**Comment by sebastienros** [comment]
Sounds great, please provide a PR.

What are you basing your performance tests on? Which code changes, and which scenarios.

**Comment by Xceno** [comment]
Hey folks, any update on this?
And a question: What about ctor parameters, can we use them in some way?

**Comment by ViRuSTriNiTy** [comment]
@Xceno 

> Okay i was a bit too quick with this one. It's actually quite easy to initialze the shape!
Just create the viewModel and pass it as an unnamend parameter to the shapeFactory.

...but this approach is exactly the thing that causes the performance issues described in #7073.

**Comment by ViRuSTriNiTy** [comment]
Ok, i have tested this again and the issue still exists.

I've tried to provide a PR by incorporating the following code

    public class BaseShapeViewModel : Orchard.DisplayManagement.Shapes.Shape
    {
      // public
        // This override ensures that the dynamic property assignment also affects the declared properties, 
        // see https://github.com/OrchardCMS/Orchard/issues/5514
        public override bool TrySetIndex(System.Dynamic.SetIndexBinder aBinder, object[] aIndexes, object aValue)
        {
          if (aIndexes.Length > 0)
          {
            var lType = this.GetType();
            System.Collections.Generic.Dictionary<string, System.Reflection.MethodInfo> lSetMethodMap;

            if (!mSetterMap.TryGetValue(lType.GUID, out lSetMethodMap))
            {
              lSetMethodMap = new System.Collections.Generic.Dictionary<string, System.Reflection.MethodInfo>();

              mSetterMap.Add(lType.GUID, lSetMethodMap); 
            }
            
            bool lResult = true;

            foreach (object lIndex in aIndexes)
            {
              string s = lIndex as string;
              if (s != null)
              {
                System.Reflection.MethodInfo lSetMethod;

                if (lSetMethodMap.TryGetValue(s, out lSetMethod))
                {  
                  lSetMethod.Invoke(this, new object[] {aValue});

                  continue;
                }

                var lProperty = lType.GetProperty(s);
                if (lProperty != null)
                {
                  if ((lProperty.PropertyType.IsPublic || lProperty.PropertyType.IsNestedPublic) && 

                    // exclude base properties as we 
                    //  a) don't want to mess with Orchard properties
                    //  b) assign values to declared properties only
                    !IsBaseProperty(lProperty) &&

                    lProperty.CanWrite) 
                  {
                    lSetMethod = lProperty.SetMethod;

                    lSetMethod.Invoke(this, new object[] {aValue});

                    lSetMethodMap.Add(s, lSetMethod);
                  }
                  else
                    lResult = lResult && base.TrySetIndex(aBinder, aIndexes, aValue);
                }
                else
                  lResult = lResult && base.TrySetIndex(aBinder, aIndexes, aValue);
              }
              else
                lResult = lResult && base.TrySetIndex(aBinder, aIndexes, aValue);
            }

            return lResult;
          }
          else
            return base.TrySetIndex(aBinder, aIndexes, aValue);
        }

        // This property allows adding dynamic properties when creating shapes in views with strongly typed syntax, e.g. @Html.Table(...)
        public new System.Collections.IDictionary Properties
        {
          get 
          {
            return base.Properties;
          }

          set
          {
            dynamic lThis = this;

            foreach (object k in value.Keys)
              lThis[k] = value[k];
          }
        }

      // private
        private bool IsBaseProperty(System.Reflection.PropertyInfo aPropertyInfo)
        {
          System.Type lBaseType = typeof(BaseShapeViewModel);
          System.Type lDeclaringType = aPropertyInfo.DeclaringType;

          while (lDeclaringType.GUID != lBaseType.GUID)
          {
            lBaseType = lBaseType.BaseType;

            if (lBaseType == null)
              return false;          
          }

          return true;
        }

        private static System.Collections.Generic.Dictionary<System.Guid, System.Collections.Generic.Dictionary<string, System.Reflection.MethodInfo>> mSetterMap =
          new System.Collections.Generic.Dictionary<System.Guid, System.Collections.Generic.Dictionary<string, System.Reflection.MethodInfo>>();
    }

into `Shape.cs` but then get a strange exception:

Cannot perform runtime binding on a null reference (   at Orchard.DisplayManagement.Descriptors.ShapeAttributeStrategy.ShapeAttributeBindingStrategy.PerformInvoke(DisplayContext displayContext, MethodInfo methodInfo, Object serviceInstance) in C:\Project\Web\Intranet.NET (Stage)\src\Orchard\DisplayManagement\Descriptors\ShapeAttributeStrategy\ShapeAttributeBindingStrategy.cs:line 73

What might be the problem here?
